### PR TITLE
Removal of unused code and tidying

### PR DESCRIPTION
**Pull request by drewnoakes**
Changes proposed in this pull request:
- Several code clean ups I've been piling together for a while.
- Biggest change is removal of around 450 LOC from `GitUICommands` through merging overloads and removing unused methods.
- Lots more unused code removed.
- Some perf and correctness fixes.

I know this is a biggie. I could split it into multiple PRs if preferred, though it'd just be chunking it into N pieces as I've squashed heavily and can't reorder commits more than this without undue pain.

What did I do to test the code and ensure quality:
- Manual review and testing

Has been tested on:
- GIT 2.16.2
- Windows 10

### CODE REVIEW

File: GitUI/CommandsDialogs/SettingsDialog/Pages/ScriptsSettingsPage.cs [hunk 1]
```diff
@@ -141,7 +141,7 @@ private void RefreshScriptDetails()
                 return;
             }
 
-            ScriptInfo scriptInfo = ScriptList.SelectedRows[0].DataBoundItem as ScriptInfo;
+            var scriptInfo = (ScriptInfo)ScriptList.SelectedRows[0].DataBoundItem;
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/ScriptsSettingsPage.cs]: I suppose absent of null check on the following line, this makes this change functionally equal - they both will fail. though the old will fail with NRE, the new will fail with invalid type cast.

File: GitUI/Plugin/PluginRegistry.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+using GitUIPluginInterfaces;
+using GitUIPluginInterfaces.RepositoryHosts;
+
+namespace GitUI
+{
+    public static class PluginRegistry
+    {
+        public static IList<IGitPlugin> Plugins { get; } = new List<IGitPlugin>();
+
+        public static List<IRepositoryHostPlugin> GitHosters { get; } = new List<IRepositoryHostPlugin>();
```
[the remark this answers is not in the document]
Comment (sharwell) [reply GitUI/Plugin/PluginRegistry.cs]: :memo: These will be going away anyway with the switch to vs-mef.

File: GitUI/CommandsDialogs/SettingsDialog/Pages/ScriptsSettingsPage.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/ScriptsSettingsPage.cs]: As @RussKie says, a cast exception gives more information than a NRE. If you are debugging without PDBs/line numbers, and know there's an exception in a method, you'll have a much better chance of correctly deducing the cause if it's a cast exception than an NRE.

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 3]
```diff
@@ -197,7 +197,7 @@ public FormBrowse(GitUICommands commands, string filter)
 
                 try
                 {
-                    PluginLoader.Load();
+                    PluginRegistry.Initialize();
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitUI/CommandsDialogs/FormBrowse.cs]: Yes, I consolidated three types related to plugins into one, and renamed 'load' to 'initialise' as you're only supposed to call it once.

**Comment by drewnoakes** [comment]
> I find it hard to believe they aren't used in some shape or form...

Hard to argue with the compiler on this one :)

Agree that this might suggest a broken pattern or some incorrect assumptions. In such cases I like to strip everything back and re-think what's needed. Removing all those events and merging so many overloads felt good. You can start to see the forest through the trees.

The ideas put forward in #4643 around splitting vertically vs. horizontally could apply equally here in future.

File: GitUI/GitUICommands.cs [hunk 4]
```diff
@@ -1215,32 +1215,29 @@ public bool StartRemotesDialog()
             return StartRemotesDialog(null);
         }
 
-        private bool StartRebaseDialog(IWin32Window owner, string onto, bool interactive = false,
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitUI/GitUICommands.cs]: The problem here was that the interface defined one set of members, but actually no one really used those members. Then the class defined the members code _actually_ wanted as overloads. And so there was a lot of duplication of signatures, and the interface prohibited directly using default parameter values. Removing the unused interface members unlocked the opportunity to consolidate the overload soup. Unfortunately it leaves a confusing diff. I don't know a way around that other than making it into dozens of commits.

File: GitUI/Plugin/PluginRegistry.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/Plugin/PluginRegistry.cs]: all good then

**Comment by RussKie** [comment]
Haha indeed. In full support of the direction

**Comment by gerhardol** [comment]
> >  I find it hard to believe they aren't used in some shape or form...
> Hard to argue with the compiler on this one :)

I believe the implementation prepared hooks for all Git operations, but not all were propagated.
They will just have to be readded if needed

**Comment by jbialobr** [comment]
GitUIPluginInterfaces.IGitUICommands events were prepared for external plugins providers. We can not know whether they are used or not.

**Comment by RussKie** [comment]
Our plugins don't use them.
We can't know about the external ones obviously, but do they exist at all?
Or are they just as mythical or significant as our Mono user base?

My vote is to remove these now for v3.
This will allow us to rethink the approach, and we can add new hooks in the
future if necessary.

V2.5x will continue to function. Plugin providers (if any) will need to
release new versions for v3.

**Comment by drewnoakes** [comment]
Agree that it's reasonable to require plugins to update for 3.0.

This whole pattern feels like it might be better suited to formalisation through a type. I haven't thought too deeply about this, but the following is the kind of thing I'm talking about.

```c#
public sealed class GitCommand<TParams>
{
    public event EventHandler<BeforeGitCommandArgs<TParams>> Before;
    public event EventHandler<AfterGitCommandArgs<TParams>> After;

    public GitOperation(
        Func<TParams, bool> handler,
        bool requiresValidWorkingDir,
        bool changesRepo)
    { ... }

    public void Invoke(TParams parameters, IWin32Window owner = null)
    { ... }
}
```

The API could be simpler if we had variadic generics.

I would like to understand more about the current use of these events and command methods before advocating for this though.

**Comment by RussKie** [comment]
Nice, I'd tweak it a bit

```
public event EventHandler<CommandExecutingEventArgs<TParams>> /* cancellable */ CommandExecuting;
public event EventHandler<CommandExecutedEventArgs<TParams>> CommandExecuted;
```

**Comment by RussKie** [comment]
Any last thoughts?

**Comment by gerhardol** [comment]
> Any last thoughts?

I did not double check all changes, but I am fine with merging

**Pull request merged**